desc/compute/utils.py

```diff
@@ -0,0 +1,303 @@
+        unique_idx = grid.unique_rho_idx
+        ds = grid.spacing[:, 1:].prod(axis=1)
+    elif surface_label == "theta":
+        surface_label_nodes = grid.nodes[:, 1]
+        unique_idx = grid.unique_theta_idx
+        ds = grid.spacing[:, [0, 2]].prod(axis=1)
+    elif surface_label == "zeta":
+        surface_label_nodes = grid.nodes[:, 2]
+        unique_idx = grid.unique_zeta_idx
+        ds = grid.spacing[:, :2].prod(axis=1)
+    else:
+        raise ValueError("Surface label must be 'rho', 'theta', or 'zeta'.")
+
+    return surface_label_nodes, unique_idx, ds
+
+
+def compress(grid, x, surface_label="rho"):
+    """
+    Compress the array x by returning only the elements in x at unique surface_label indices.
+
+    Parameters
+    ----------
+    grid : Grid, LinearGrid, ConcentricGrid, QuadratureGrid
+        Collocation grid containing the nodes to evaluate at.
+    x : ndarray
+        The array to compress.
+        Should usually represent a surface function (a function constant over a surface)
+        in an array that matches the grid's pattern.
+    surface_label : str
+        The surface label of rho, theta, or zeta.
+
+    Returns
+    -------
+    ndarray
+        x[grid.unique_surface_label_indices]
+        This array will be sorted such that the
+            first element corresponds to the value associated with the smallest surface
+            last element  corresponds to the value associated with the largest surface
+    """
+    if surface_label == "rho":
+        return x[grid.unique_rho_idx]
+    if surface_label == "theta":
+        return x[grid.unique_theta_idx]
+    if surface_label == "zeta":
+        return x[grid.unique_zeta_idx]
+    raise ValueError("Surface label must be 'rho', 'theta', or 'zeta'.")
+
+
+def expand(grid, x, surface_label="rho"):
+    """
+    Expand the array x by duplicating elements in x to match the grid's pattern.
+
+    Parameters
+    ----------
+    grid : LinearGrid, ConcentricGrid, QuadratureGrid
+        Collocation grid containing the nodes to evaluate at.
+        Note that support for concentric grids on surface_label="theta" is not implemented yet.
+    x : ndarray
+        Stores the values of a surface function (a function constant over a surface)
+        for all unique surfaces of the specified label on the grid.
+        - len(x) should be grid.num_surface_label
+        - x should be sorted such that the
+            first element corresponds to the value associated with the smallest surface
+            last element  corresponds to the value associated with the largest surface
+    surface_label : str
+        The surface label of rho, theta, or zeta.
+
+    Returns
+    -------
+    ndarray
+        X expanded to match the grid's pattern.
+    """
+    # TODO: This function assumes the grid is sorted like a typical grid.
+    #   If we could store the inverse array when we call np.unique() in grid.py, then this function would become
```

Comment:
seems like that might be the easier case, i dont see any reason why we can't store the inverse indices as an attribute of the grid